### ISSUE TITLE
Threading Issue

### ISSUE THREAD

**Issue by nillario** [chezou/tabula-py#204]
# Summary of your issue
I've been using tabula to scrape some tables from pdfs online, and ran into some file/permission issues with the threading aspect. 

I've been drawing from the this post: 
https://chriskiehl.com/article/parallelism-in-one-line
to parallelize my scraping of tables, and I found that I sometimes ran into file/permission errors. I was running this on Windows 10, and my error was often usually something like:
```
FileNotFoundError: [WinError 2] The system cannot find the file specified: '3856.pdf' or
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: '24764.pdf'
```
My understanding is that tabula is generating these temporary pdfs to read them, but is somehow running into issues when tabula is called in different places at the same time. My understanding of threading isn't great, so I'm not exactly sure what the underlying interaction is. 
I've been able to reproduce this with many pdfs, so in the code here I just used the 17 page Twain book, A Dog's Tale. This pdf has no tables, but I was receiving the same problems on some table-laden pdfs and I don't think that's the issue.

# Check list before submit

- [x] Did you read [FAQ](https://tabula-py.readthedocs.io/en/latest/faq.html)
 I couldn't find anything in the FAQ, but it's very possible I don't know what I'm looking for.

- [x] (Optional, but really helpful) Your PDF URL: https://www.freeclassicebooks.com/Mark%20Twain/A%20Horse's%20Tale.pdf

- [x] Paste the output of `import tabula; tabula.environment_info()` on Python REPL: 

```Python version:
    3.7.4 (tags/v3.7.4:e09359112e, Jul  8 2019, 20:34:20) [MSC v.1916 64 bit (AMD64)]
Java version:
    java version "1.8.0_231"

Java(TM) SE Runtime Environment (build 1.8.0_231-b11)

Java HotSpot(TM) 64-Bit Server VM (build 25.231-b11, mixed mode)
tabula-py version: 1.4.3
platform: Windows-10-10.0.18362-SP0
uname:
    uname_result(system='Windows', node='Kolya-PC', release='10', version='10.0.18362', machine='AMD64', processor='Intel64 Family 6 Model 94 Stepping 3, GenuineIntel')
linux_distribution: ('MINGW32_NT-6.2', '1.0.19', '')
mac_ver: ('', ('', '', ''), '')
```

If not possible to execute `tabula.environment_info()`, please answer following questions manually.

- [ ] Paste the output of `python --version` command on your terminal: ?
- [ ] Paste the output of `java -version` command on your terminal: ?
- [ ] Does `java -h` command work well?; Ensure your java command is included in `PATH`
- [ ] Write your OS and it's version: ?

# What did you do when you faced the problem?
I haven't worked out a solution yet but I think I'm going to try locking the read_pdf function.

## Code:

```
import numpy as np
import pandas as pd
import requests
import tabula
import io
from multiprocessing.dummy import Pool as ThreadPool
dogs_tale = "https://www.freeclassicebooks.com/Mark%20Twain/A%20Dog's%20Tale.pdf"


def test_func(url):
    page = requests.get(url, stream=True)
    with io.BytesIO(page.content) as fh:
        df_list = tabula.read_pdf(input_path=fh, pages = "all", multiple_tables=True, stream=True)
    print(len(df_list))

def scrape_manual(times_scraped):
    pool = ThreadPool(3)
    results = pool.map(test_func, [dogs_tale for i in range(times_scraped)])
    pool.close()
    pool.join()

if __name__ == "__main__":
    scrape_manual(10)
```

## Expected behavior:

This function has minimal expected output, but shouldn't error, (which it does most of the time).

```
0
0
0
0
0
0
0
0
0
```

## Actual behavior:

```
0
0
0
0
Traceback (most recent call last):
  File "C:\Users\Kolya\Documents\Work\tabula_test.py", line 23, in <module>
    scrape_manual(10)
  File "C:\Users\Kolya\Documents\Work\tabula_test.py", line 18, in scrape_manual
    results = pool.map(test_func, [dogs_tale for i in range(times_scraped)])
  File "C:\Users\Kolya\AppData\Local\Programs\Python\Python37\lib\multiprocessing\pool.py", line 268, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "C:\Users\Kolya\AppData\Local\Programs\Python\Python37\lib\multiprocessing\pool.py", line 657, in get
    raise self._value
  File "C:\Users\Kolya\AppData\Local\Programs\Python\Python37\lib\multiprocessing\pool.py", line 121, in worker
    result = (True, func(*args, **kwds))
  File "C:\Users\Kolya\AppData\Local\Programs\Python\Python37\lib\multiprocessing\pool.py", line 44, in mapstar
    return list(map(*args))
  File "C:\Users\Kolya\Documents\Work\tabula_test.py", line 13, in test_func
    df_list = tabula.read_pdf(input_path=fh, pages = "all", multiple_tables=True, stream=True)
  File "C:\Users\Kolya\AppData\Local\Programs\Python\Python37\lib\site-packages\tabula\wrapper.py", line 295, in read_pdf
    os.unlink(path)
FileNotFoundError: [WinError 2] The system cannot find the file specified: '21976.pdf'
```

## Related Issues:
None I could immediately find.

**Comment by chezou** [comment]
@nillario Thanks for reporting. This issue should be resolved by #205 
I'm releasing v2.0.2 on PyPI. Can you try it out?

**Comment by nillario** [comment]
Looks all fixed! I had managed to fix my problem by using a new function
```
def tabula_lock(*args, **kwargs):
    with lock:
        ret_obj = tabula.read_pdf(*args, **kwargs)
    return ret_obj
```
but this isn't needed with the new fix, which seems structurally better.
